app/src/ui/lib/text-box.tsx

```diff
@@ -143,21 +110,29 @@ export class TextBox extends React.Component<ITextBoxProps, ITextBoxState> {
     }
   }
 
+  public componentWillReceiveProps(nextProps: ITextBoxProps) {
+    if (this.state.value !== nextProps.value) {
+      this.setState({ value: nextProps.value })
+    }
+  }
+
   private onChange = (event: React.FormEvent<HTMLInputElement>) => {
-    this.storeCaretPosition(event.currentTarget)
+    const value = event.currentTarget.value
 
     if (this.props.onChange) {
       this.props.onChange(event)
     }
 
-    if (this.props.onValueChanged && !event.defaultPrevented) {
-      this.props.onValueChanged(event.currentTarget.value)
-    }
+    const defaultPrevented = event.defaultPrevented
+
+    this.setState({ value }, () => {
+      if (this.props.onValueChanged && !defaultPrevented) {
```

Comment:
Doing this in the callback to `setState` means that we know the state has been applied _before_ we raise the callback (and potentially trigger `componentWillReceiveProps` again).

Reply:
TIL `setState` accepts a callback function.